Add subsection "Default Parameters for Overloaded Methods" (#2646)
scala/docs.scala-lang: _tour/default-parameter-values.md

```
{% endtab %}
{% endtabs %}

### Default Parameters for Overloaded Methods

> It's important to notice that Scala doesn't allow having two methods with default parameters and with the same name (overloaded). The reason is to keep a deterministic naming-scheme for the generated methods which return default arguments.

Reviewer: > The reason is to keep a deterministic naming-scheme for the generated methods which return default arguments.

This is an implementation detail which could change in the future

```suggestion
It's important to notice that Scala doesn't allow having two methods with default parameters and with the same name (overloaded).
```

> For example, if you write `def f(a: Int = 1)`, the compiler generates `def f$default$1 = 1`. If you have two overloads with defaults on the same parameter position, then we would need a different naming scheme. In this situation, it is preferred to keep the generated byte-code stable over multiple compiler runs.

Author: Sorry, which line would you remove? Is the example too related with a low-level design choice that may change in the future? If so I can remove the second paragraph.

> For example, if you write `def f(a: Int = 1)`, the compiler generates `def f$default$1 = 1`. If you have two overloads with defaults on the same parameter position, then we would need a different naming scheme. In this situation, it is preferred to keep the generated byte-code stable over multiple compiler runs.

Reviewer: > For example, if you write `def f(a: Int = 1)`, the compiler generates `def f$default$1 = 1`. If you have two overloads with defaults on the same parameter position, then we would need a different naming scheme. In this situation, it is preferred to keep the generated byte-code stable over multiple compiler runs.

Sorry yes to be clear, line 70 doesnt say much when you remove the low level implementation detail, so I think the code example line 77 is clear enough

> For example, if you write `def f(a: Int = 1)`, the compiler generates `def f$default$1 = 1`. If you have two overloads with defaults on the same parameter position, then we would need a different naming scheme. In this situation, it is preferred to keep the generated byte-code stable over multiple compiler runs.

Author: Got it! Removed that part.

> The other reason is to avoid the ambiguity that can be caused due to the existence of default parameters. To illustrate the problem, let's consider the method declarations provided below:

Reviewer: ```suggestion
An important reason why is to avoid the ambiguity that can be caused due to the existence of default parameters. To illustrate the problem, let's consider the method declarations provided below:
```

{% tabs default-parameter-values-5 %}
{% tab 'Scala 2 and 3' for=default-parameter-values-5 %}

> ```scala 

Reviewer: This will check in the CI that the snippet fails to compile
```suggestion
```scala mdoc:fail
```

object A {
  def func(x: Int = 34): Unit
  def func(y: String = "abc"): Unit
}
```
{% endtab %}

> {% endtabs %}

Reviewer: This final change is just for consistency
~~~suggestion
{% tabs default-parameter-values-5 class=tabs-scala-version %}
{% tab 'Scala 2' %}
```scala mdoc:fail
object A {
  def func(x: Int = 34): Unit
  def func(y: String = "abc"): Unit
}
```
{% endtab %}
{% tab 'Scala 3' %}
```scala
object A:
  def func(x: Int = 34): Unit
  def func(y: String = "abc"): Unit
```
{% endtab %}
{% endtabs %}
~~~
